### PR TITLE
Introduce Vitest

### PR DESCRIPTION
**Pull request by hl662**
Continuing off of the work from #6890 

For `core-bentley`,`core-geometry`, `core-common`:
- Refactors all tests
- Uses vitest for coverage as well, needs explicit `@vitest/coverage-istanbul` (I'm open to switching to node's native code coverage `v8` too, no opinions on either)
- Removed explicit `chai`, `chai-as-promised`, `mocha`, `nyc` dependencies

OneAtATime tests in core-bentley are skipped for now, haven't finished refactoring them.

Switching the integration tests to vitest is out of the scope of this PR, so just looking at the unit tests within each sub repo instead. We'll investigate this more after.

A couple more issues to circle back to later on as a result of this: #7200 #7201

### CODE REVIEW

File: core/geometry/package.json [hunk 1]
```diff
@@ -4,7 +4,7 @@
   "description": "iTwin.js Core Geometry library",
   "main": "lib/cjs/core-geometry.js",
   "module": "lib/esm/core-geometry.js",
-  "typings": "lib/cjs/core-geometry",
+  "typings": "lib/esm/core-geometry",
```
Comment (aruniverse) [inline core/geometry/package.json:7]: dont forget to revert

File: core/geometry/src/test/setupTests.ts [hunk 2]
```diff
@@ -0,0 +1 @@
+import "../core-geometry"; // Needed to avoid circular import runtime errors with vitest
```
Comment (aruniverse) [inline core/geometry/src/test/setupTests.ts:1]: please add a comment with a link to the vitest docs you found this in

File: full-stack-tests/core/webpack.config.js [hunk 3]
```diff
@@ -47,6 +47,10 @@ function createConfig(shouldInstrument) {
         path: require.resolve("path-browserify"),
         stream: require.resolve("stream-browserify"),
         zlib: require.resolve("browserify-zlib"),
+        tty: require.resolve("tty-browserify"),
```
Comment (aruniverse) [inline full-stack-tests/core/webpack.config.js:50]: where are these additions coming from?

File: .vscode/settings.json [hunk 4]
```diff
@@ -7,7 +7,7 @@
     "source.fixAll.markdownlint": "explicit"
   },
   "editor.insertSpaces": true,
-  "editor.formatOnSave": true,
+  // "editor.formatOnSave": true,
```
Comment (aruniverse) [inline .vscode/settings.json:10]: dont forget to revert this file

File: full-stack-tests/core/webpack.config.js (hunk 3 above)
Comment (hl662) [reply full-stack-tests/core/webpack.config.js:50]: [error.txt](https://github.com/user-attachments/files/17138427/error.txt)
See the above for the failures in particular, I saw these errors pop up when I picked up your old PR, I assumed it was related to changing the typings, but after reverting the typings changes it still stuck around...

File: .vscode/settings.json (hunk 4 above)
Comment (hl662) [reply .vscode/settings.json]: done

File: core/geometry/src/test/setupTests.ts (hunk 2 above)
Comment (hl662) [reply core/geometry/src/test/setupTests.ts:3]: added

File: core/geometry/package.json (hunk 1 above)
Comment (hl662) [reply core/geometry/package.json]: done

File: core/common/vitest.config.mts [hunk 5]
```diff
@@ -0,0 +1,12 @@
+import { defineConfig } from 'vitest/config';
+export default defineConfig({
+  test: {
+    dir: "src",
+    coverage: {
+      provider: "istanbul", // akin to nyc,
```
[the remark this answers is not in the document]
Comment (hl662) [reply core/common/vitest.config.mts:6]: agh, forgot to add that, just added

File: core/bentley/package.json [hunk 6]
```diff
@@ -4,22 +4,22 @@
   "description": "Bentley JavaScript core components",
   "main": "lib/cjs/core-bentley.js",
   "module": "lib/esm/core-bentley.js",
-  "typings": "lib/cjs/core-bentley",
+  "typings": "lib/esm/core-bentley",
```
[the remark this answers is not in the document]
Comment (hl662) [reply core/bentley/package.json]: done

File: full-stack-tests/core/webpack.config.js (hunk 3 above)
Comment (aruniverse) [reply full-stack-tests/core/webpack.config.js:50]: well looking at the errors, prob coming from a dependency update, which is fine...

File: core/common/src/test/FeatureSymbology.test.ts [hunk 7]
```diff
@@ -84,42 +84,39 @@ describe("FeatureAppearance", () => {
...
-        const app = FeatureAppearance.fromJSON(appProps);
-        expect(app.viewDependentTransparency).to.equal(expected);
-        expect(app.toJSON().viewDependentTransparency).to.equal(expected);
-      }
-
-      test({ }, false);
-      test({ transparency: undefined }, false);
-      test({ transparency: 1 }, false);
-      test({ transparency: 0 }, false );
+  it("view-dependent transparency serialization", () => {
```
[the remark this answers is not in the document]
Comment (hl662) [reply core/common/src/test/FeatureSymbology.test.ts:87]: nested describes can work, but it doesn't seem necessary - it doesn't make use of any test hooks, there's only one test case nested inside the outer one, so I figured let's just move them out and coalesce into just one test case

File: core/geometry/package.json [hunk 8]
```diff
@@ -39,28 +39,16 @@
...
     "rimraf": "^3.0.2",
-    "typescript": "~5.3.3"
+    "typescript": "~5.3.3",
+    "vitest": "^2.1.0"
   },
   "dependencies": {
     "@itwin/core-bentley": "workspace:*",
     "flatbuffers": "~1.12.0"
-  },
-  "nyc": {
```
Comment (hl662) [inline core/geometry/package.json:58]: @aruniverse @saeeedtorabi did this work in the past? I copied the functionality of the coverage below into geometry's vitest.config.mts and it's now failing coverage, saying `Coverage for functions (53.06%) does not meet global threshold (85%)`

File: .vscode/settings.json [hunk 9]
```diff
@@ -104,7 +100,5 @@
     "NODE_ENV": "development"
   },
   "typescript.tsdk": "common/temp/node_modules/typescript/lib",
-  "cSpell.words": [
-    "queryables"
-  ],
-}
\ No newline at end of file
+  "cSpell.words": ["queryables"]
```
Comment (aruniverse) [inline .vscode/settings.json:103]: move this to the cspell file, and revert changes to this file since they all come from prettier formatter

File: core/bentley/package.json [hunk 10]
```diff
@@ -12,14 +12,14 @@
     "directory": "core/bentley"
   },
   "scripts": {
-    "build": "npm run -s build:cjs && npm run -s build:esm",
+    "build": "npm run -s build:esm && npm run -s build:cjs",
```
Comment (aruniverse) [inline core/bentley/package.json:15]: lets revert this change

File: core/geometry/package.json (hunk 8 above)
Comment (saeeedtorabi) [reply core/geometry/package.json:58]: Currently on the master branch our coverage for `core-geometry` is 
```
Statements   : 93.53% ( 38252/40895 )
Branches     : 82.55% ( 15344/18587 )
Functions    : 92.47% ( 6318/6832 )
Lines        : 93.53% ( 36824/39368 )
```

@dassaf4 Any idea why @hl662  is getting 53% coverage?!

File: core/bentley/src/test/OneAtATime.test.ts [hunk 11]
```diff
@@ -3,20 +3,21 @@
...
-const expect = chai.expect;
+// const assert = chai.assert;
+// const expect = chai.expect;
 
-chai.use(chaiAsPromised);
+// chai.use(chaiAsPromised);
 describe("OneAtATime test", () => {
 
-  it("OneAtATime", async () => {
+  it.skip("OneAtATime", async () => {
```
Comment (aruniverse) [inline core/bentley/src/test/OneAtATime.test.ts:20]: idk if i ever finished refactoring this test

File: core/geometry/package.json (hunk 8 above)
Comment (hl662) [reply core/geometry/package.json:58]: I'll check for regressions - maybe the one in master excludes certain file paths

File: core/bentley/src/test/OneAtATime.test.ts (hunk 11 above)
Comment (hl662) [reply core/bentley/src/test/OneAtATime.test.ts:20]: This is a tricky one I skipped for now as well, I need to put this in the PR description

File: core/geometry/package.json (hunk 8 above)
Comment (hl662) [reply core/geometry/package.json:58]: vitest's `expect` has different syntaxes than chai's `expect`, perhaps when we use the old syntax, that isn't caught by the coverage - I'll refactor all of the `expect` methods and see if it helps

Comment (aruniverse) [reply core/geometry/package.json:58]: take a look at `tools/build/.nycrc`

Comment (hl662) [reply core/geometry/package.json:58]: none on the test files - I'm looking at the local coverage report right now, there's no `test.ts`

Comment (dassaf4) [reply core/geometry/package.json:58]: I see only one core-geometry test file with the `expect` calls changed to `expectTo`. I assume this is an experiment and the goal is to match the current coverage percentages with the new test framework. Otherwise, core-geometry changes look good so far. Let us know if you need help @hl662!

File: core/bentley/src/test/OneAtATime.test.ts (hunk 11 above)
Comment (aruniverse) [reply core/bentley/src/test/OneAtATime.test.ts:20]: make a todo in the pr description, and create issues so these things dont get lost

File: core/geometry/package.json (hunk 8 above)
Comment (hl662) [reply core/geometry/package.json:58]: image
Update: I updated the vitest config coverage's `include` and `exclude` and got the numbers above, we might be missing some covered lines and statements by a small fraction, lmk what you think @saeeedtorabi @dassaf4

File: core/bentley/src/test/OneAtATime.test.ts (hunk 11 above)
Comment (hl662) [reply core/bentley/src/test/OneAtATime.test.ts:20]: Issue opened: https://github.com/iTwin/itwinjs-core/issues/7200

File: core/bentley/package.json (hunk 10 above)
Comment (hl662) [reply core/bentley/package.json]: done

File: .vscode/settings.json (hunk 9 above)
Comment (hl662) [reply .vscode/settings.json]: moved, and reverted the other formatting

**Review by aruniverse — APPROVED** [review]
approving on behalf of core bentley & common

File: core/geometry/src/test/clipping/ClipPrimitives.test.ts [hunk 12]
```diff
@@ -320,7 +320,7 @@ describe("ClipPrimitive", () => {
   let polygonB: Point3d[];
   let polygonC: Point3d[];
 
-  before(() => {
+  beforeEach(() => {
```
[the remark this answers is not in the document]
Comment (hl662) [reply core/geometry/src/test/clipping/ClipPrimitives.test.ts:323]: yes, I'll fix that, thanks!

**Review by eringram — APPROVED** [review]

File: core/geometry/src/test/clipping/ClipPrimitives.test.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (eringram) [reply core/geometry/src/test/clipping/ClipPrimitives.test.ts:323]: Np. Possibly the same change applies to ClipVector.test.ts

File: core/geometry/package.json (hunk 8 above)
Comment (dassaf4) [reply core/geometry/package.json:58]: Sweet! I'm fine with those numbers. These are estimates of estimates, so the decimal differences are unimportant. Besides, according to L. Kronecker, "God created the integers, all else is the work of man." ;)

File: core/geometry/src/test/clipping/ClipPrimitives.test.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (dassaf4) [reply core/geometry/src/test/clipping/ClipPrimitives.test.ts:323]: Good catch @eringram. I think there were other `beforeEach` changes in core-geom...

[the remark this answers is not in the document]
Comment (hl662) [reply core/geometry/src/test/clipping/ClipPrimitives.test.ts:323]: I checked there were only those 2 - vitest **needs** things to switch from `before` -> `beforeAll` and `after` -> `afterAll`. `beforeEach` and `afterEach` stays the same between vitest and mocha. If there were any other beforeEach imports that were missing, we would have caught it at build time.

**Review by dassaf4 — APPROVED** [review]

File: core/geometry/src/test/clipping/ClipPrimitives.test.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (hl662) [reply core/geometry/src/test/clipping/ClipPrimitives.test.ts]: okay, fixed both

File: core/geometry/src/test/geometry3d/Angle.test.ts [hunk 13]
```diff
@@ -129,7 +130,7 @@ describe("Angle.AlmostEqual", () => {
     ck.testFalse(Angle.createRadians(1.0e-20).isExactZero);
 
     ck.checkpoint("End Angle.AlmostEqual");
-    expect(ck.getNumErrors()).equals(0);
+    expect(ck.getNumErrors()).toEqual(0);
```
Comment (dassaf4) [inline core/geometry/src/test/geometry3d/Angle.test.ts:133]: Are these equals -> toEqual changes still needed?

Comment (hl662) [reply core/geometry/src/test/geometry3d/Angle.test.ts:133]: not necessarily, functionally both behave the same, but I was going to leave it in there as an example of switching over from chai to vitest's testing syntax. chai, out of the box, doesn't support asynchronous assertions like how vitest does. The author of vitest [recommends](https://github.com/vitest-dev/vitest/issues/2843#issuecomment-1424352509) to not mix and match, but if we want to continue using the assertions API, we can continue importing that from within vitest. I can remove it this, to avoid inconsistency

Comment (dassaf4) [reply core/geometry/src/test/geometry3d/Angle.test.ts:133]: I guess we could use some best practices for vitest going forward. What is the difference between `expect(1).equals(0)` and `expect(1).toEqual(0)`?

Comment (hl662) [reply core/geometry/src/test/geometry3d/Angle.test.ts:133]: For primitive types, they are the exact same, but objects are more complex. In vitest, when comparing the expected value to the actual value, you can choose between toBe() or toEqual(). The former is strict equality (think ===). The latter **deeply**, recursively checks the equality of all property fields, not the object's identity. [Reference](https://stackoverflow.com/questions/45195025/what-is-the-difference-between-tobe-and-toequal-in-jest). Chai's .[equals](https://www.chaijs.com/api/bdd/#method_equal)() (or `eq`) would check an object is strictly equal, while for deeply equals you would use `eql`. Unfamiliar people would not be able to tell as easily in chai, that `equal` and `eql` behave differently, whereas vitest tries to be more explicit in their method naming.

Comment (dassaf4) [reply core/geometry/src/test/geometry3d/Angle.test.ts:133]: Ah, I see. Nice explanation! All of the occurrences of ".equals" in core-geometry operate on numbers or booleans. So to stick with vitest syntax, it would be more natural to switch them to `.toBe`. If you want to do this, feel free. Otherwise, we'll do it.

**Review by ben-polinsky — APPROVED** [review]

File: core/geometry/src/test/geometry3d/Angle.test.ts (hunk 13 above)
Comment (hl662) [reply core/geometry/src/test/geometry3d/Angle.test.ts:133]: Yes, toBe would be appropriate in those cases. I'll defer to you to update the syntax, I'll be giving a demo internally next week on some things to catch when refactoring

Comment (aruniverse) [reply core/geometry/src/test/geometry3d/Angle.test.ts:133]: David can we do that in a separate follow up pr then?

Comment (dassaf4) [reply core/geometry/src/test/geometry3d/Angle.test.ts:133]: Looking forward to your demo!

**Comment by aruniverse** [comment]
> BTW, just pulled and tried out vitest, and the multithreading is _so nice_. All core-geometry tests done in 50s. Wow.
- @dassaf4 
- 
just in case anyone comes back to this pr, dont want you to miss this comment :)
